expression/builtin_control.go

```diff
@@ -48,47 +53,312 @@ type caseWhenFunctionClass struct {
+	fieldTps := make([]*types.FieldType, 0, (l+1)/2)
+	decimal, flen, isBinaryStr := args[1].GetType().Decimal, 0, false
+	for i := 1; i < l; i += 2 {
+		fieldTps = append(fieldTps, args[i].GetType())
+		decimal = mathutil.Max(decimal, args[i].GetType().Decimal)
+		flen = mathutil.Max(flen, args[i].GetType().Flen)
+		isBinaryStr = isBinaryStr || types.IsBinaryStr(args[i].GetType())
+	}
+	if l%2 == 1 {
+		fieldTps = append(fieldTps, args[l-1].GetType())
+		decimal = mathutil.Max(decimal, args[l-1].GetType().Decimal)
+		flen = mathutil.Max(flen, args[l-1].GetType().Flen)
+		isBinaryStr = isBinaryStr || types.IsBinaryStr(args[l-1].GetType())
+	}
+	fieldTp := types.AggFieldType(fieldTps)
+	classType := types.AggTypeClass(fieldTps, &fieldTp.Flag)
+	fieldTp.Decimal, fieldTp.Flen = decimal, flen
+	var tp evalTp
+	switch classType {
+	case types.ClassInt:
+		tp = tpInt
+		fieldTp.Decimal = 0
+	case types.ClassReal:
+		tp = tpReal
+	case types.ClassDecimal:
+		tp = tpDecimal
+	case types.ClassString:
+		tp = tpString
+		if !isBinaryStr {
+			fieldTp.Charset, fieldTp.Collate = mysql.DefaultCharset, mysql.DefaultCollationName
+		}
+		if types.IsTypeTime(fieldTp.Tp) {
+			tp = tpTime
+		} else if fieldTp.Tp == mysql.TypeDuration {
+			tp = tpDuration
+		}
+	}
+	// Set retType to BINARY(0) if all arguments are of type NULL.
+	if fieldTp.Tp == mysql.TypeNull {
+		fieldTp.Flen, fieldTp.Decimal = 0, -1
+		types.SetBinChsClnFlag(fieldTp)
+	}
+	argTps := make([]evalTp, 0, l)
+	for i := 0; i < l-1; i += 2 {
+		argTps = append(argTps, tpInt, tp)
+	}
+	if l%2 == 1 {
+		argTps = append(argTps, tp)
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tp, argTps...)
+	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinCaseWhenSig{newBaseBuiltinFunc(args, ctx)}
+	bf.tp = fieldTp
+	switch classType {
+	case types.ClassInt:
+		sig = &builtinCaseWhenIntSig{baseIntBuiltinFunc{bf}}
+	case types.ClassReal:
+		sig = &builtinCaseWhenRealSig{baseRealBuiltinFunc{bf}}
+	case types.ClassDecimal:
+		sig = &builtinCaseWhenDecimalSig{baseDecimalBuiltinFunc{bf}}
+	case types.ClassString:
+		sig = &builtinCaseWhenStringSig{baseStringBuiltinFunc{bf}}
+		if types.IsTypeTime(fieldTp.Tp) {
+			sig = &builtinCaseWhenTimeSig{baseTimeBuiltinFunc{bf}}
+		} else if fieldTp.Tp == mysql.TypeDuration {
+			sig = &builtinCaseWhenDurationSig{baseDurationBuiltinFunc{bf}}
+		}
+	}
 	return sig.setSelf(sig), nil
 }
 
-type builtinCaseWhenSig struct {
-	baseBuiltinFunc
+type builtinCaseWhenIntSig struct {
+	baseIntBuiltinFunc
 }
 
-// eval evals a builtinCaseWhenSig.
+// evalInt evals a builtinCaseWhenIntSig.
 // See https://dev.mysql.com/doc/refman/5.7/en/case.html
-func (b *builtinCaseWhenSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return types.Datum{}, errors.Trace(err)
+func (b *builtinCaseWhenIntSig) evalInt(row []types.Datum) (int64, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	var (
+		condition, ret int64
```

Comment:
move `ret, isNull, and err` to the return statement?
`func (b *builtinCaseWhenIntSig) evalInt(row []types.Datum) (ret int64, isNull bool, err error) `